cmd/syncthing/gui.go

```diff
@@ -1024,6 +1027,96 @@ func (s *apiService) getSystemLogTxt(w http.ResponseWriter, r *http.Request) {
+	name string
+	data []byte
+}
+
+func (s *apiService) getSupportBundle(w http.ResponseWriter, r *http.Request) {
+	var arrayFile []fileEntry
+
+	// Redacted configuration as a JSON
+	jsonConfig, _ := json.MarshalIndent(getRedactedConfig(s), "", "  ")
+	arrayFile = append(arrayFile, fileEntry{name: "config.json", data: jsonConfig})
+
+	// Log as a text
+	var buff bytes.Buffer
+	for _, line := range s.systemLog.Since(time.Time{}) {
+		fmt.Fprintf(&buff, "%s: %s\n", line.When.Format(time.RFC3339), line.Message)
+	}
+	arrayFile = append(arrayFile, fileEntry{name: "log.txt", data: buff.Bytes()})
+
+	// Errors as a JSON
+	if errs := s.guiErrors.Since(time.Time{}); len(errs) > 0 {
+		jsonError, _ := json.MarshalIndent(errs, "", "  ")
+		arrayFile = append(arrayFile, fileEntry{name: "errors.json", data: jsonError})
+	}
+
+	// Errors as a JSON
+	panicFiles, ok := checkPanicFile(baseDirs["config"])
+	if ok {
+		for _, f := range panicFiles {
+			panicF, err := ioutil.ReadFile(f)
+			if err != nil {
+				arrayFile = append(arrayFile, fileEntry{name: filepath.Base(f), data: panicF})
+			}
+		}
+	}
+
+	// Version and platform information as a JSON
+	versionPlatform, _ := json.MarshalIndent(map[string]string{
+		"version":     Version,
+		"codename":    Codename,
+		"longVersion": LongVersion,
+		"os":          runtime.GOOS,
+		"arch":        runtime.GOARCH,
+	}, "", "  ")
+	arrayFile = append(arrayFile, fileEntry{name: "versionPlatform.json", data: versionPlatform})
+
+	// Report Data as a JSON
+	version := usageReportVersion
+	if val, _ := strconv.Atoi(r.URL.Query().Get("version")); val > 0 {
+		version = val
+	}
+
+	usageReportingData, _ := json.MarshalIndent(reportData(s.cfg, s.model, s.connectionsService, version, true), "", "  ")
+	arrayFile = append(arrayFile, fileEntry{name: "usageReportingData.json", data: usageReportingData})
+
+	// Heap and CPU Proofs as a pprof extension
+	var heapBuffer, cpuBuffer bytes.Buffer
+	filename := fmt.Sprintf("syncthing-heap-%s-%s-%s-%s.pprof", runtime.GOOS, runtime.GOARCH, Version, time.Now().Format("150405")) // hhmmss
+	runtime.GC()
+	pprof.WriteHeapProfile(&heapBuffer)
+	arrayFile = append(arrayFile, fileEntry{name: filename, data: heapBuffer.Bytes()})
+
+	const duration = 4 * time.Second
+	filename = fmt.Sprintf("syncthing-cpu-%s-%s-%s-%s.pprof", runtime.GOOS, runtime.GOARCH, Version, time.Now().Format("150405")) // hhmmss
+	pprof.StartCPUProfile(&cpuBuffer)
+	time.Sleep(duration)
+	pprof.StopCPUProfile()
+	arrayFile = append(arrayFile, fileEntry{name: filename, data: cpuBuffer.Bytes()})
+
+	path := filepath.Join(baseDirs["config"], "support_bundle")
+	os.MkdirAll(path, 0700)
+
+	output := time.Now().Format("2006-01-02T15.04.05") + ".zip"
+	newfilepath := filepath.Join(path, output)
+	newfile, _ := os.Create(newfilepath)
+	defer newfile.Close()
+
+	// Add files to zip
+	zipBuffer(newfile, arrayFile)
+
+	// Delete old zips over 10 (fifo)
+	eraseSupportBundleZip(path)
+
+	from, _ := os.Open(newfilepath)
```

Comment:
You're doing something odd here, I think maybe serving the file twice. `http.ServeFile` serves directly from the filesystem so no need to open first. (You do however need to close the writing side, above.) Alternately we could do this in memory - have `zipBuffer` zip *into* another buffer, write those to file and also write the same bytes to the request.

Reply:
@calmh Thanks your comment. Yes, I recognized that the file served twice. File opening and `io.Copy(w, from)` part is unnecessary. But, sorry that I didn't understand `(You do however need to close the writing side, above.)` part of ur comment. I appreciate if u explain it.

Reply:
If you write the zip file to an in memory bytes.Buffer this is a non-issue. But if you, like now, write it to an actual file on disk and then read it back, you have the problem that until the writing side has `Close()`d the file, all contents might not be written and hence cannot be read yet. For zip files this is definitely the case because the close call is what writes the directory at the end of the file. Before calling close, the zip file is not a zip file.

Reply:
Ohh I got it, thanks for explanation.